pkg/quarterdeck/auth.go

```diff
@@ -637,6 +639,126 @@ func (s *Server) Refresh(c *gin.Context) {
 	c.JSON(http.StatusOK, out)
 }
 
+// Switch re-authenticates users (human users only) using the access token the user
+// posts in the headers in order to give the user new claims with a new organization ID.
+// E.g. the user switches from being logged into one organization to being logged into
+// another organization. The user must submit the orgID of the organization they wish
+// to switch to and the user must belong to that organization otherwise an error is
+// returned.
+//
+// NOTE: this endpoint cannot be used with api keys because api keys are only ever
+// issued to one organization (and in fact, one project inside of one organization).
+// Only human users can belong to multiple organizations.
+func (s *Server) Switch(c *gin.Context) {
+	var (
+		err error
+		in  *api.SwitchRequest
+		out *api.LoginReply
+	)
+
+	if err = c.BindJSON(&in); err != nil {
+		sentry.Warn(c).Err(err).Msg("could not parse refresh request")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse(api.ErrUnparsable))
+		metrics.FailedLogins.WithLabelValues(ServiceName, UserHuman, "unparseable").Inc()
+		return
+	}
+
+	// Ensure the orgID is included in the request
+	if ulids.IsZero(in.OrgID) {
+		log.Debug().Msg("missing orgID in switch request")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("missing organization id"))
+		metrics.FailedLogins.WithLabelValues(ServiceName, UserHuman, "missing org_id").Inc()
+		return
+	}
+
+	// Parse the claims from the access token in the request
+	var claims *tokens.Claims
+	if claims, err = middleware.GetClaims(c); err != nil {
+		sentry.Error(c).Err(err).Msg("could not get user claims from authenticated request")
+		c.JSON(http.StatusUnauthorized, api.ErrorResponse(api.ErrInvalidUserClaims))
+		metrics.FailedLogins.WithLabelValues(ServiceName, UserHuman, "no access token").Inc()
+		return
+	}
+
+	// If the orgID in the claims is the same as the requested orgID return an error
+	// (the user must use the refresh endpoint to get claims for the same orgID)
+	if orgID := claims.ParseOrgID(); orgID.Compare(in.OrgID) == 0 {
+		sentry.Warn(c).Msg("user attempting to switch to the same organization")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("cannot switch into the organization you are currently logged into"))
+		metrics.FailedLogins.WithLabelValues(ServiceName, UserHuman, "switch to same org").Inc()
+		return
+	}
```

Comment:
I debated this - I would hate for the `switch` endpoint to be used for refreshing tokens; but this would require the front-end to filter the list of organizations so that the user couldn't click on the organization they're already logged into. Any thoughts on this check?

Reply:
The frontend has access to the orgID in the claims right? So my hope is that filtering the list isn't too difficult, and I think it would be potentially confusing for the user to see their current org in the list if the option is called "switch". Preventing the switch to the same organization seems like the right behavior to me.

Reply:
Awesome; then we'll leave this in. Would you mind helping me make sure that Edwin and Cletus understand this functionality?